internal/router/middlewares/ratelim.go

```diff
@@ -77,3 +81,62 @@ func extractClientIP(r *http.Request) (string, error) {
 	}
 	return ip, nil
 }
+
+type middleware struct {
+	store   limiter.Store
+	keyFunc httplimit.KeyFunc
+
+	// clients with key are not affected by rate limiter
+	apiKey string
+}
+
+// Handle returns the HTTP handler as a middleware. This handler calls Take() on
+// the store and sets the common rate limiting headers. If the take is
+// successful, the remaining middleware is called. If take is unsuccessful, the
+// middleware chain is halted and the function renders a 429 to the caller with
+// metadata about when it's safe to retry.
+func (m *middleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		// Call the key function - if this fails, it's an internal server error.
+		key, err := m.keyFunc(r)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// skip rate limiting checks if secret key is provided
+		if key := r.Header.Get("Secret-Key"); key != "" && m.apiKey != "" {
```

Comment:
should we name the header "Api-Key" instead of "Secret-Key" to make it consistent with the code?

Reply:
hmm maybe i get your point now. this is the special "API key" that can bypass the checks. so it makes sense.